File: app/servicos/comprovante_epi.py
"""O comprovante de entrega de EPI: contexto congelado e impressão em .docx.

Fatia 2 do módulo Gestão de EPI. Está aqui a mesma divisão que o parecer e o
certificado já têm: este arquivo guarda o **contexto** e o **render** (como
`documento.py` e `certificado.py`), e `epi_ficha.py` guarda as regras de negócio
(como `parecer.py` e `emissao_certificado.py`).

**Por que existe um documento, e por que ele é papel.** A decisão 8 (18/08/2026)
fixou que o comprovante de EPI é papel assinado no ato e digitalizado depois:
assinatura em tela tem valor probatório mais fraco e exigiria dispositivo no
almoxarifado, e transformar cada par de luva num documento do SEI não se
sustenta. O sistema imprime, o servidor assina, e o PDF volta como
`Anexo(categoria='FICHA_EPI', assinado=True)`.

**O que o documento diz é o que valia no dia da entrega.** `montar_contexto` tem
uma única porta de saída para registro gravado, e ela lê `contexto_congelado` —
nunca o catálogo de hoje. É a RN-15 aplicada ao EPI: renomear "Luva nitrílica"
para "Luva de procedimento" em 2027 não pode reescrever uma entrega de 2024.

**E o documento não carrega a data em que foi impresso**, de propósito. A
segunda via de um comprovante de 2024 tem de sair com texto idêntico ao da
primeira — mesma disciplina do texto de ouro do parecer —, e um carimbo de
"emitido em" faria cada reimpressão divergir da anterior por construção. Quem
imprimiu e quando fica na trilha de auditoria, que é onde essa informação
pertence.

O modelo (`comprovante_epi_v1.docx`) é produzido por
`ferramentas/gerar_modelo_comprovante_epi.py` e pode ser substituído por um
desenhado no Word, desde que use os mesmos marcadores — a mesma liberdade que o
modelo do parecer tem.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from pathlib import Path

from app.config import obter_config
from app.servicos import datas_br, documento, textos

MODELO = "comprovante_epi_v1.docx"
SUBPASTA = "epi"
TIPO_DOCX = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"

# O termo que o servidor assina. Fica aqui, em código, e não em `texto_padrao`:
# ele é a razão de ser do papel — sem as obrigações declaradas, o que se colhe é
# um recibo de entrega, não um termo de responsabilidade.
#
# A citação é à **NR-6 como um todo**, sem número de item, e isso é deliberado.
# O conteúdo das obrigações é o da norma; o dispositivo exato ainda não foi
# conferido contra a redação vigente, e há uma questão aberta antes dela — as NRs
# são instrumento da CLT e o titular aqui é servidor estatutário, o que está
# registrado como pendência no `docs/ROPA.md` §9, item 1. Citar item e alínea que
# eu não conferi seria dar ao papel uma precisão que ele não tem.
TERMO_RESPONSABILIDADE = (
    "Declaro ter recebido o equipamento de proteção individual acima "
    "discriminado, em perfeitas condições de uso, e ter sido orientado quanto "
    "ao seu uso, guarda e conservação.\n"
    "Comprometo-me, nos termos da NR-6, a: usar o equipamento apenas para a "
    "finalidade a que se destina; responsabilizar-me por sua guarda e "
    "conservação; comunicar ao setor qualquer alteração que o torne impróprio "
    "para uso; e cumprir as determinações sobre o uso adequado.\n"
    "Estou ciente de que a devolução, a troca por desgaste, o extravio ou a "
    "danificação devem ser comunicados à CSSO, e de que o equipamento é de uso "
    "pessoal e intransferível."
)

ROTULO_TIPO = {
    "ENTREGA": "COMPROVANTE DE ENTREGA DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
    "DEVOLUCAO": "COMPROVANTE DE DEVOLUÇÃO DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
    "SUBSTITUICAO": "COMPROVANTE DE SUBSTITUIÇÃO DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
    "DESCARTE": "COMPROVANTE DE DESCARTE DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
    "ESTORNO": "ESTORNO DE REGISTRO DA FICHA DE EPI",
}


@dataclass
class ContextoComprovante:
    """Tudo o que o comprovante imprime — e nada que dependa do banco de hoje.

    Todo campo é `str`, `int` ou `date`. Não há FK, não há objeto do ORM e não há
    consulta: depois de congelado, este contexto sobrevive a renomear o item, a
    inativar o lote, a transferir o servidor de unidade e a desativar a conta de
    quem entregou.
    """

    # --- identificação do registro ---
    # `registro_id` NÃO entra no congelado, e isso é uma consequência direta da
    # trava do banco: o id só existe depois do INSERT, e gravar o congelado
    # depois seria alterar uma coluna de conteúdo — que a tabela recusa. Também
    # não faria sentido congelá-lo: é a chave primária da própria linha, o único
    # dado que não pode divergir dela. `descongelar` o recebe de volta da linha.
    registro_id: int
    tipo: str
    data_evento: date
    quantidade: int
    unidade_medida: str

    # --- quem recebeu, como estava no dia ---
    servidor_nome: str
    siape: str
    cargo: str
    funcao: str
    unidade: str
    posto: str

    # --- o que foi entregue, como estava no dia ---
    epi_nome: str
    categoria: str
    fabricante: str
    modelo: str
    normas: str
    numero_ca: str
    validade_ca: date | None
    lote: str
    tamanho: str
    previsao_troca: date | None

    # --- rastreabilidade da compra pública ---
    pregao: str
    item_pregao: str
    empenho: str
    nota_fiscal: str
    fornecedor: str

    # --- quem entregou e por quem responde ---
    entregue_por: str
    motivo: str

    # --- cabeçalho institucional ---
    setor_sigla: str
    setor_nome: str
    setor_endereco: str
    cidade: str

    termo: str = TERMO_RESPONSABILIDADE

    CAMPOS_CONGELADOS = (
        "tipo", "data_evento", "quantidade", "unidade_medida",
        "servidor_nome", "siape", "cargo", "funcao", "unidade", "posto",
        "epi_nome", "categoria", "fabricante", "modelo", "normas", "numero_ca",
        "validade_ca", "lote", "tamanho", "previsao_troca",
        "pregao", "item_pregao", "empenho", "nota_fiscal", "fornecedor",
        "entregue_por", "motivo",
        "setor_sigla", "setor_nome", "setor_endereco", "cidade", "termo",
    )

    # as três datas viram texto ISO no congelado e voltam a `date` no descongelar
    _DATAS = ("data_evento", "validade_ca", "previsao_troca")

    def congelar(self) -> dict:
        """A forma que vai para `epi_ficha_registro.contexto_congelado`.

        Data vira texto ISO explicitamente, e não pela conversão implícita do
        `JSONTexto`: a coluna serializa com `default=str`, então um `date` iria
        e voltaria como texto de qualquer modo — mas em silêncio, e
        `descongelar` teria de adivinhar quais chaves eram datas. Escrito aqui,
        o par congelar/descongelar é simétrico e verificável.
        """
        dados = {campo: getattr(self, campo) for campo in self.CAMPOS_CONGELADOS}
        for campo in self._DATAS:
            valor = dados[campo]
            dados[campo] = valor.isoformat() if valor is not None else None
        return dados

    @classmethod
    def descongelar(cls, dados: dict, *, registro_id: int) -> "ContextoComprovante":
        valores = {campo: dados.get(campo) for campo in cls.CAMPOS_CONGELADOS}
        for campo in cls._DATAS:
            bruto = valores.get(campo)
            valores[campo] = date.fromisoformat(bruto[:10]) if bruto else None
        return cls(registro_id=registro_id, **valores)

    # -----------------------------------------------------------------
    @property
    def titulo(self) -> str:
        return ROTULO_TIPO.get(self.tipo, ROTULO_TIPO["ENTREGA"])

    @property
    def identificacao_do_equipamento(self) -> str:
        """Fabricante, marca e modelo numa linha só, sem travessão órfão."""
        partes = [p for p in (self.fabricante, self.modelo) if p]
        return " · ".join(partes) or "—"

    @property
    def aquisicao(self) -> str:
        """Pregão, item, empenho e nota fiscal — o que amarra o EPI à compra.

        Vazio quando a entrega não veio de lote (entrega de balcão): imprimir
        "Pregão: — · Empenho: —" ocuparia uma linha para dizer nada.
        """
        rotulados = (
            ("Pregão", self.pregao),
            ("Item", self.item_pregao),
            ("Empenho", self.empenho),
            ("Nota fiscal", self.nota_fiscal),
            ("Fornecedor", self.fornecedor),
        )
        presentes = [f"{rotulo}: {valor}" for rotulo, valor in rotulados if valor]
        return " · ".join(presentes)

    def como_dicionario(self) -> dict:
        def data(valor: date | None) -> str:
            return datas_br.numerica(valor) if valor else "—"

        return {
            "titulo": self.titulo,
            "registro": f"{self.registro_id:06d}",
            "setor_sigla": self.setor_sigla,
            "setor_nome": self.setor_nome,
            "setor_endereco": self.setor_endereco,
            "cidade": self.cidade,
            "servidor_nome": self.servidor_nome,
            "siape": self.siape,
            "cargo": self.cargo or "—",
            "funcao": self.funcao or "—",
            "unidade": self.unidade or "—",
            "posto": self.posto or "—",
            "epi_nome": self.epi_nome,
            "categoria": self.categoria,
            "identificacao": self.identificacao_do_equipamento,
            # RichText: norma de EPI vem uma por linha, dentro da mesma célula —
            # o mesmo `_rich` que o parecer usa nos postos de trabalho
            "normas": documento.rich_multilinha(self.normas) or "—",
            # o CA é o que constitui o equipamento como EPI (NR-6): sai em
            # destaque e com a validade ao lado, nunca só o número
            "numero_ca": self.numero_ca or "não se aplica",
            "validade_ca": data(self.validade_ca),
            "lote": self.lote or "—",
            "tamanho": self.tamanho or "—",
            "quantidade": f"{self.quantidade} {self.unidade_medida.lower()}",
            "data_evento": data(self.data_evento),
            "data_extenso": datas_br.por_extenso(self.data_evento),
            "previsao_troca": data(self.previsao_troca),
            "aquisicao": self.aquisicao or "—",
            "entregue_por": self.entregue_por,
            "motivo": self.motivo or "—",
            "termo": documento.rich_multilinha(self.termo) or "",
        }

    def obrigatorios_faltantes(self) -> list[str]:
        exigidos = {
            "servidor_nome": "nome do servidor",
            "siape": "matrícula SIAPE",
            "epi_nome": "nome do EPI",
            "categoria": "categoria da NR-6",
            "entregue_por": "quem entregou",
        }
        return [rotulo for campo, rotulo in exigidos.items() if not getattr(self, campo)]


def caminho_saida(registro_id: int, ano: int) -> Path:
    """`dados/documentos/epi/{ano}/Comprovante_EPI_000123.docx`.

    Por ano, como o parecer, o certificado e a lista de presença, porque é assim
    que o setor arquiva e é assim que o backup separa.
    """
    cfg = obter_config()
    pasta = cfg.caminho(cfg.dir_documentos) / SUBPASTA / str(ano)
    return pasta / f"Comprovante_EPI_{registro_id:06d}.docx"


def renderizar(contexto: ContextoComprovante, destino: Path) -> dict:
    """Passa pelo mecanismo único de documento do sistema.

    Dois mecanismos de documento seriam dois lugares para o hash de integridade
    divergir, e o parecer já provou que um só basta.
    """
    return documento.renderizar_modelo(
        MODELO,
        contexto.como_dicionario(),
        destino,
        dica="Rode python -m ferramentas.gerar_modelo_comprovante_epi.",
    )


def nome_para_download(contexto: ContextoComprovante) -> str:
    """`Comprovante_EPI_000123_Marco_Antonio.docx` — legível na pasta de quem baixa."""
    partes = [
        "Comprovante_EPI",
        f"{contexto.registro_id:06d}",
        textos.slug_ascii(contexto.servidor_nome)[:60],
    ]
    return "_".join(p for p in partes if p) + ".docx"


__all__ = [
    "MODELO",
    "ROTULO_TIPO",
    "SUBPASTA",
    "TERMO_RESPONSABILIDADE",
    "TIPO_DOCX",
    "ContextoComprovante",
    "caminho_saida",
    "nome_para_download",
    "renderizar",
]

File: nuvem/src/servicos/comprovante_epi.ts
/**
 * O comprovante de entrega de EPI: contexto congelado e impressão em .docx.
 * Porte de `app/servicos/comprovante_epi.py`.
 *
 * Mesma divisão do parecer e do certificado: aqui o **contexto** e o
 * **render**; em `epi_ficha.ts`, as regras.
 *
 * **Por que é papel.** A decisão 8 (18/08/2026) fixou que o comprovante é papel
 * assinado no ato e digitalizado depois. O sistema imprime, o servidor assina,
 * e o PDF volta como `Anexo(categoria='FICHA_EPI', assinado=true)`.
 *
 * **O que o documento diz é o que valia no dia da entrega** (RN-15): a única
 * porta de saída para registro gravado lê `contexto_congelado`, nunca o
 * catálogo de hoje. E o documento não carrega a data em que foi impresso: a
 * segunda via tem de sair com texto idêntico ao da primeira.
 *
 * Na nuvem o .docx vai para o armazenamento (`armazenamento.ts`), e não para
 * uma pasta local; PDF não há (PORTE.md §8) — o comprovante já era .docx.
 */
import * as datas_br from "./datas_br.js";
import * as documento from "./documento.js";
import * as textos from "./textos.js";

export const MODELO = "comprovante_epi_v1.docx";
export const SUBPASTA = "epi";
export const TIPO_DOCX = "application/vnd.openxmlformats-officedocument.wordprocessingml.document";

// O termo que o servidor assina. Fica em código: sem as obrigações declaradas o
// que se colhe é um recibo, não um termo de responsabilidade. A citação é à
// NR-6 como um todo, sem item — o dispositivo exato ainda não foi conferido
// (ROPA §9, item 1).
export const TERMO_RESPONSABILIDADE =
  "Declaro ter recebido o equipamento de proteção individual acima " +
  "discriminado, em perfeitas condições de uso, e ter sido orientado quanto " +
  "ao seu uso, guarda e conservação.\n" +
  "Comprometo-me, nos termos da NR-6, a: usar o equipamento apenas para a " +
  "finalidade a que se destina; responsabilizar-me por sua guarda e " +
  "conservação; comunicar ao setor qualquer alteração que o torne impróprio " +
  "para uso; e cumprir as determinações sobre o uso adequado.\n" +
  "Estou ciente de que a devolução, a troca por desgaste, o extravio ou a " +
  "danificação devem ser comunicados à CSSO, e de que o equipamento é de uso " +
  "pessoal e intransferível.";

export const ROTULO_TIPO: Record<string, string> = {
  ENTREGA: "COMPROVANTE DE ENTREGA DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
  DEVOLUCAO: "COMPROVANTE DE DEVOLUÇÃO DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
  SUBSTITUICAO: "COMPROVANTE DE SUBSTITUIÇÃO DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
  DESCARTE: "COMPROVANTE DE DESCARTE DE EQUIPAMENTO DE PROTEÇÃO INDIVIDUAL",
  ESTORNO: "ESTORNO DE REGISTRO DA FICHA DE EPI",
};

export const CAMPOS_CONGELADOS = [
  "tipo", "data_evento", "quantidade", "unidade_medida",
  "servidor_nome", "siape", "cargo", "funcao", "unidade", "posto",
  "epi_nome", "categoria", "fabricante", "modelo", "normas", "numero_ca",
  "validade_ca", "lote", "tamanho", "previsao_troca",
  "pregao", "item_pregao", "empenho", "nota_fiscal", "fornecedor",
  "entregue_por", "motivo",
  "setor_sigla", "setor_nome", "setor_endereco", "cidade", "termo",
] as const;

// as três datas: texto ISO no congelado, 'AAAA-MM-DD' aqui também
const DATAS = ["data_evento", "validade_ca", "previsao_troca"] as const;

export interface DadosComprovante {
  registro_id: number;
  tipo: string;
  data_evento: string;
  quantidade: number;
  unidade_medida: string;
  servidor_nome: string;
  siape: string;
  cargo: string;
  funcao: string;
  unidade: string;
  posto: string;
  epi_nome: string;
  categoria: string;
  fabricante: string;
  modelo: string;
  normas: string;
  numero_ca: string;
  validade_ca: string | null;
  lote: string;
  tamanho: string;
  previsao_troca: string | null;
  pregao: string;
  item_pregao: string;
  empenho: string;
  nota_fiscal: string;
  fornecedor: string;
  entregue_por: string;
  motivo: string;
  setor_sigla: string;
  setor_nome: string;
  setor_endereco: string;
  cidade: string;
  termo?: string;
}

/**
 * Tudo o que o comprovante imprime — e nada que dependa do banco de hoje.
 *
 * `registro_id` NÃO entra no congelado: o id só existe depois do INSERT, e
 * gravar o congelado depois seria alterar coluna de conteúdo — que a trava
 * recusa. `descongelar` o recebe de volta da linha.
 */
export class ContextoComprovante {
  registro_id!: number;
  tipo!: string;
  data_evento!: string;
  quantidade!: number;
  unidade_medida!: string;
  servidor_nome!: string;
  siape!: string;
  cargo!: string;
  funcao!: string;
  unidade!: string;
  posto!: string;
  epi_nome!: string;
  categoria!: string;
  fabricante!: string;
  modelo!: string;
  normas!: string;
  numero_ca!: string;
  validade_ca!: string | null;
  lote!: string;
  tamanho!: string;
  previsao_troca!: string | null;
  pregao!: string;
  item_pregao!: string;
  empenho!: string;
  nota_fiscal!: string;
  fornecedor!: string;
  entregue_por!: string;
  motivo!: string;
  setor_sigla!: string;
  setor_nome!: string;
  setor_endereco!: string;
  cidade!: string;
  termo: string = TERMO_RESPONSABILIDADE;

  constructor(d: DadosComprovante) {
    Object.assign(this, d);
    if (d.termo === undefined || d.termo === null) this.termo = TERMO_RESPONSABILIDADE;
  }

  /** A forma que vai para `epi_ficha_registro.contexto_congelado`. */
  congelar(): Record<string, unknown> {
    const dados: Record<string, unknown> = {};
    for (const campo of CAMPOS_CONGELADOS) dados[campo] = (this as any)[campo] ?? null;
    for (const campo of DATAS) {
      const v = dados[campo] as string | null;
      dados[campo] = v ? v.slice(0, 10) : null;
    }
    return dados;
  }

  static descongelar(dados: Record<string, unknown>, opcoes: { registro_id: number }): ContextoComprovante {
    const valores: Record<string, unknown> = {};
    for (const campo of CAMPOS_CONGELADOS) valores[campo] = dados[campo] ?? null;
    for (const campo of DATAS) {
      const bruto = valores[campo] as string | null;
      valores[campo] = bruto ? String(bruto).slice(0, 10) : null;
    }
    return new ContextoComprovante({ registro_id: opcoes.registro_id, ...(valores as any) });
  }

  get titulo(): string {
    return ROTULO_TIPO[this.tipo] ?? ROTULO_TIPO["ENTREGA"]!;
  }

  /** Fabricante, marca e modelo numa linha só, sem travessão órfão. */
  get identificacao_do_equipamento(): string {
    return [this.fabricante, this.modelo].filter((p) => p).join(" · ") || "—";
  }

  /** Pregão, item, empenho e nota fiscal — vazio na entrega de balcão. */
  get aquisicao(): string {
    const rotulados: [string, string][] = [
      ["Pregão", this.pregao],
      ["Item", this.item_pregao],
      ["Empenho", this.empenho],
      ["Nota fiscal", this.nota_fiscal],
      ["Fornecedor", this.fornecedor],
    ];
    return rotulados
      .filter(([, valor]) => valor)
      .map(([rotulo, valor]) => `${rotulo}: ${valor}`)
      .join(" · ");
  }

  como_dicionario(): Record<string, unknown> {
    const data = (valor: string | null) => (valor ? datas_br.numerica(valor) : "—");
    return {
      titulo: this.titulo,
      registro: String(this.registro_id).padStart(6, "0"),
      setor_sigla: this.setor_sigla,
      setor_nome: this.setor_nome,
      setor_endereco: this.setor_endereco,
      cidade: this.cidade,
      servidor_nome: this.servidor_nome,
      siape: this.siape,
      cargo: this.cargo || "—",
      funcao: this.funcao || "—",
      unidade: this.unidade || "—",
      posto: this.posto || "—",
      epi_nome: this.epi_nome,
      categoria: this.categoria,
      identificacao: this.identificacao_do_equipamento,
      // norma de EPI vem uma por linha, dentro da mesma célula
      normas: documento.rich_multilinha(this.normas) ?? "—",
      // o CA é o que constitui o equipamento como EPI (NR-6)
      numero_ca: this.numero_ca || "não se aplica",
      validade_ca: data(this.validade_ca),
      lote: this.lote || "—",
      tamanho: this.tamanho || "—",
      quantidade: `${this.quantidade} ${(this.unidade_medida ?? "").toLowerCase()}`,
      data_evento: data(this.data_evento),
      data_extenso: datas_br.por_extenso(this.data_evento),
      previsao_troca: data(this.previsao_troca),
      aquisicao: this.aquisicao || "—",
      entregue_por: this.entregue_por,
      motivo: this.motivo || "—",
      termo: documento.rich_multilinha(this.termo) ?? "",
    };
  }

  obrigatorios_faltantes(): string[] {
    const exigidos: [keyof ContextoComprovante, string][] = [
      ["servidor_nome", "nome do servidor"],
      ["siape", "matrícula SIAPE"],
      ["epi_nome", "nome do EPI"],
      ["categoria", "categoria da NR-6"],
      ["entregue_por", "quem entregou"],
    ];
    return exigidos.filter(([campo]) => !this[campo]).map(([, rotulo]) => rotulo);
  }
}

/**
 * A chave no armazenamento: `documentos/epi/{ano}/Comprovante_EPI_000123.docx`.
 * Por ano, como o parecer e o certificado.
 */
export function caminho_saida(registro_id: number, ano: number): string {
  return `documentos/${SUBPASTA}/${ano}/Comprovante_EPI_${String(registro_id).padStart(6, "0")}.docx`;
}

/** Passa pelo mecanismo único de documento do sistema. */
export function renderizar(contexto: ContextoComprovante, destino: string | null) {
  return documento.renderizar_modelo(MODELO, contexto.como_dicionario(), {
    destino,
    dica: "Rode ferramentas/converter-modelos-docx.ts.",
  });
}

/** `Comprovante_EPI_000123_Marco_Antonio.docx`. */
export function nome_para_download(contexto: ContextoComprovante): string {
  const partes = [
    "Comprovante_EPI",
    String(contexto.registro_id).padStart(6, "0"),
    textos.slug_ascii(contexto.servidor_nome).slice(0, 60),
  ];
  return partes.filter((p) => p).join("_") + ".docx";
}
